### PR TITLE
Update warning warn_explicit

### PR DESCRIPTION
**Pull request by oow214**
Update warning module https://github.com/RustPython/RustPython/issues/4013

- [x] normalize_module
- [x] already_warned
- [x] get_filter

### CODE REVIEW

File: vm/src/warn.rs [hunk 1]
```diff
@@ -31,19 +31,91 @@ pub fn warn(
...
-    // TODO: Implement correctly
+    // if registry.is(vm.ctx.types.none_type.as_object()) || !registry.dict() {
+    //     return Err(vm.new_type_error("'registry' must be a dict or None".to_string()));
+    // }
+
+    let module = if module.is_some() {
+        normalize_module(filename, vm)
+    } else {
+        module
+    };
```
Comment (TwoPair) [inline vm/src/warn.rs:103]:
```suggestion
    let module = module.unwrap_or(normalize_module(filename, vm));
```
I think you can make it simpler like this.

**Comment by youknowone** [comment]
The checklist is looking ambitious! We will finally reach to the goal, but you don't need to finish all of them in single PR. Whenever you want to cut the PR in preferred size and make it merged, you can consider to move the checklist to the issue and finish a PR as a part of the work.

File: vm/src/warn.rs [hunk 2]
```diff
@@ -31,19 +32,119 @@ pub fn warn(
...
+    let module = module.and(normalize_module(filename, vm));
+
+    /* module can be None if a warning is emitted late during Python shutdown.
+    In this case, the Python warnings module was probably unloaded, filters
+    are no more available to choose as action. It is safer to ignore the
+    warning and do nothing. */
+    if module.is_none() {
+        // return Ok(&vm.ctx.types.none_type);
+        return Ok(());
+    }
```
Comment (youknowone) [inline vm/src/warn.rs:141]:
```suggestion
    let module = if let Some(module) = module {
        module
    } else {
        // return Ok(&vm.ctx.types.none_type);
        return Ok(());
    }
```

To let typesystem recognize module never be None later.

File: vm/src/warn.rs [hunk 3]
```diff
@@ -31,19 +62,132 @@ pub fn warn(
...
+    _lineno: usize,
+    _module: PyObjectRef,
+    mut _item: PyObjectRef,
+    vm: &VirtualMachine,
+) -> PyResult<()> {
+    let filters = vm.state.warnings.filters.as_object().to_owned();
+    let filters = PyListRef::try_from_object(vm, filters);
+    if filters.is_err() {
+        return Err(vm.new_value_error("_warnings.filters must be a list".to_string()));
+    }
```
Comment (youknowone) [inline vm/src/warn.rs:77]:
```suggestion
    let filters: PyListRef = filters.try_into_value(vm).map_err(|_| vm.new_value_error("_warnings.filters must be a list".to_string()))?;
```

File: vm/src/warn.rs [hunk 4]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+    filename: PyStrRef,
+    lineno: usize,
+    module: Option<PyObjectRef>,
+    registry: PyObjectRef,
     _source_line: Option<PyObjectRef>,
     _source: Option<PyObjectRef>,
     vm: &VirtualMachine,
 ) -> PyResult<()> {
-    // TODO: Implement correctly
+    if registry.clone().is_true(vm)?
```
Comment (Snowapril) [inline vm/src/warn.rs:169]: Is this necessary? It seems warn_explicit spec only suppose to have `dict` type or `None` for registry argument.

File: vm/src/warn.rs [hunk 5]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+        return Err(vm.new_type_error("'registry' must be a dict or None".to_string()));
+    }
+
+    // Normalize module.
+    let module = module.and(normalize_module(filename, vm));
+    let module = if let Some(module) = module {
+        module
+    } else {
+        return Ok(());
+    };
```
Comment (Snowapril) [inline vm/src/warn.rs:182]:
```suggestion
    let module = match module.or_else(|| normalize_module(filename, vm)) {
        Some(module) => module,
        None => return Ok(())
    };
```

File: vm/src/warn.rs [hunk 6]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+    module: Option<PyObjectRef>,
+    registry: PyObjectRef,
     _source_line: Option<PyObjectRef>,
     _source: Option<PyObjectRef>,
     vm: &VirtualMachine,
 ) -> PyResult<()> {
-    // TODO: Implement correctly
+    if registry.clone().is_true(vm)?
+        && !registry.class().is(vm.ctx.types.dict_type)
+        && !registry.class().is(vm.ctx.types.none_type)
```
Comment (Snowapril) [inline vm/src/warn.rs:171]:
```suggestion
        && !vm.is_none(registry.as_object())
```
We need to check with singleton

File: vm/src/warn.rs [hunk 7]
```diff
@@ -55,8 +197,35 @@ fn warn_explicit(
...
+            vm.ctx.new_str(text).into(),
+            category.as_object().to_owned(),
+            vm.ctx.new_int(lineno).into(),
+        ],
+        &vm.ctx,
+    );
+
+    if already_warned(registry, key.as_object().to_owned(), 0, vm) {
+        return Ok(());
+    }
```
Comment (Snowapril) [inline vm/src/warn.rs:214]: We only need to check already_warned when registry is not `None`.
```suggestion
    if !vm.is_none(registry.as_object()) && already_warned(registry, key.as_object().to_owned(), 0, vm) {
        return Ok(());
    }
```

File: vm/src/warn.rs [hunk 8]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+        module
+    } else {
+        return Ok(());
+    };
+
+    // Normalize message.
+    let text = message.as_str();
+    if text.is_empty() {
+        return Ok(());
+    }
```
Comment (Snowapril) [inline vm/src/warn.rs:188]: It seems text should not always be `str` type.
```suggestion
    let (category, text, message) = if message.is_instance(vm.ctx.exceptions.warning.as_object(), vm)? {
        (message.class().into_pyobject(vm), message.str(vm)?.into_pyobject(vm), message)
    } else {
        (category, message, vm.invoke(category.as_object(), (message, ))?)
    };
```
I am not sure that this is the best option. please take a look this code @youknowone

File: vm/src/warn.rs [hunk 9]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+            Err(vm.new_value_error(format!("_warnings.filters item {} isn't a 5-tuple", i)))
+        }?;
+
+        /* Python code: action, msg, cat, mod, ln = item */
+        let _action = tmp_item.get(0);
+        let _msg = tmp_item.get(1);
+        let _cat = tmp_item.get(2);
+        let _item_mod = tmp_item.get(3);
+        let _ln_obj = tmp_item.get(4);
+    }
```
Comment (Snowapril) [inline vm/src/warn.rs:99]: Don't we have to check the condition for these variables?

Comment (Snowapril) [reply vm/src/warn.rs:99]: It seems this method returns nothing actually.. And we need a default action for when there is no available one in the filter list.

File: vm/src/warn.rs [hunk 10]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+    false
+}
+
+fn normalize_module(filename: PyStrRef, vm: &VirtualMachine) -> Option<PyObjectRef> {
+    let len = filename.char_len();
+
+    if len == 0 {
+        Some(vm.new_pyobj("<unknown>"))
+    } else if len >= 3 && filename.as_str().contains(".py") {
+        Some(vm.new_pyobj(filename.as_str().replace(".py", "")))
```
Comment (Snowapril) [inline vm/src/warn.rs:151]: We need not use replace, maybe `substr` is suitable for this kind of code.

File: vm/src/warn.rs [hunk 11]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+    Ok(())
+}
+
+fn already_warned(
+    registry: PyObjectRef,
+    key: PyObjectRef,
+    should_set: usize,
+    vm: &VirtualMachine,
+) -> bool {
+    let version_obj = registry.get_item("version", vm).ok();
```
Comment (Snowapril) [inline vm/src/warn.rs:110]: We can make this "version" as interned string. Please refer `context.rs` and `identifier!` macro.

File: vm/src/warn.rs [hunk 12]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+    let filters_version = vm
+        .ctx
+        .new_int(vm.state.warnings.filters_version)
+        .as_object()
+        .to_owned();
+
+    if version_obj.is_none()
+        || version_obj.as_ref().unwrap().try_int(vm).is_err()
+        || !version_obj.unwrap().is(&filters_version)
```
Comment (Snowapril) [inline vm/src/warn.rs:119]: It seems the use of `is_none` and `unwarp` combination is very dangerous. We can replace this code to `match` stmt or other functional way.

File: vm/src/warn.rs [hunk 13]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+    {
+        let registry = registry.dict();
+        registry.as_ref().map(|registry| {
+            registry.into_iter().collect_vec().clear();
+            registry
+        });
```
Comment (Snowapril) [inline vm/src/warn.rs:125]: There is `clear` method in PyDict. We can use it after converting that method as public.

File: vm/src/warn.rs [hunk 14]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+    key: PyObjectRef,
+    should_set: usize,
+    vm: &VirtualMachine,
+) -> bool {
+    let version_obj = registry.get_item("version", vm).ok();
+    let filters_version = vm
+        .ctx
+        .new_int(vm.state.warnings.filters_version)
+        .as_object()
+        .to_owned();
```
Comment (Snowapril) [inline vm/src/warn.rs:115]:
```suggestion
    let filters_version = vm
        .ctx
        .new_int(vm.state.warnings.filters_version)
        .into_pyobject(vm);
```

File: vm/src/warn.rs [hunk 15]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+        });
+
+        if let Some(registry) = registry {
+            if registry.set_item("version", filters_version, vm).is_err() {
+                return false;
+            }
+        }
+    } else {
+        let already_warned = registry.get_item(key.as_ref(), vm);
+        return already_warned.is_ok();
```
Comment (Snowapril) [inline vm/src/warn.rs:134]:
```suggestion
        return already_warned.is_ok();
```
```suggestion
        let already_warned = registry.get_item(key.as_ref(), vm)?;
        if already_warned.is_true(vm)? {
            return Ok(true);
        }
```
We need to return when the `already_warned` is true not for `Ok` in `PyResult`

File: vm/src/warn.rs [hunk 16]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+    }
+
+    /* This warning wasn't found in the registry, set it. */
+    if should_set != 0 {
+        registry.set_item(key.as_ref(), vm.new_pyobj(""), vm).ok();
+    }
+
+    false
```
Comment (Snowapril) [inline vm/src/warn.rs:142]:
```suggestion
    Ok(if should_set != 0 {
        registry.set_item(key.as_ref(), vm.ctx.true_value, vm).map(|_| true)?;
    } else {
        false
    })
```
Where this `vm.new_pyobj("")` come from?

File: vm/src/warn.rs [hunk 17]
```diff
@@ -31,19 +62,130 @@ pub fn warn(
...
+        let _ln_obj = tmp_item.get(4);
+    }
+
+    Ok(())
+}
+
+fn already_warned(
+    registry: PyObjectRef,
+    key: PyObjectRef,
+    should_set: usize,
```
Comment (Snowapril) [inline vm/src/warn.rs:107]:
```suggestion
    should_set: bool,
```
Maybe trivial, It seems this does not have to be `usize` type.

File: vm/src/warn.rs (hunk 6 above)
Comment (youknowone) [inline vm/src/warn.rs:171]: In RustPython, we can encode it to `Option<PyDictRef>`
```rust
let registry: Option<PyDictRef> = registry.try_into_value(vm).map_err(|_| vm.new_type_error("'registry' must be a dict or None".to_owned()))?;
```

File: vm/src/warn.rs (hunk 8 above)
Comment (youknowone) [reply vm/src/warn.rs:188]: the suggestion looks good.

**Comment by oow214** [comment]
```
let _action = tmp_item.get(0);
let _msg = tmp_item.get(1);
let _cat = tmp_item.get(2);
let _item_mod = tmp_item.get(3);
let _ln_obj = tmp_item.get(4);
```
Is there a function to fetch items at once in PyTuple?

**Comment by youknowone** [comment]
Did you try `as_slice()` of `PyTuple`?

File: vm/src/warn.rs [hunk 18]
```diff
@@ -31,19 +76,171 @@ pub fn warn(
...
+        false
+    })
+}
+
+fn normalize_module(filename: PyStrRef, vm: &VirtualMachine) -> Option<PyObjectRef> {
+    let len = filename.char_len();
+
+    if len == 0 {
+        Some(vm.new_pyobj("<unknown>"))
+    } else if len >= 3 && filename.as_str().contains(".py") {
```
Comment (youknowone) [inline vm/src/warn.rs:215]:
```suggestion
    } else if len >= 3 && filename.as_str().ends_with(".py") {
```

**Comment by oow214** [comment]
There is still work to be done in warn_explicit. This will be done in a new PR.
I've edited as far as I know, can you please check if there is still work to be done?

File: vm/src/warn.rs [hunk 19]
```diff
@@ -31,19 +76,172 @@ pub fn warn(
...
+        let tmp_item = if let Some(tmp_item) = tmp_item {
+            let tmp_item = PyTupleRef::try_from_object(vm, tmp_item)?;
+            if tmp_item.len() != 5 {
+                Err(vm.new_value_error(format!("_warnings.filters item {} isn't a 5-tuple", i)))
+            } else {
+                Ok(tmp_item)
+            }
+        } else {
+            Err(vm.new_value_error(format!("_warnings.filters item {} isn't a 5-tuple", i)))
+        }?;
```
Comment (oow214) [inline vm/src/warn.rs:119]: Is there a better way to handle errors?

**Comment by oow214** [comment]
There is still work to be done in warn_explicit. This will be done in a new PR.
I've edited as far as I know, can you please check if there is still work to be done?

File: vm/src/warn.rs [hunk 20]
```diff
@@ -31,19 +78,172 @@ pub fn warn(
...
+        let tmp_item = if let Some(tmp_item) = tmp_item {
+            let tmp_item = PyTupleRef::try_from_object(vm, tmp_item)?;
+            if tmp_item.len() == 5 {
+                Ok(tmp_item)
+            } else {
+                Err(vm.new_value_error(format!("_warnings.filters item {} isn't a 5-tuple", i)))
+            }
+        } else {
+            Err(vm.new_value_error(format!("_warnings.filters item {} isn't a 5-tuple", i)))
+        }?;
```
Comment (oow214) [inline vm/src/warn.rs:121]: Is there a better way to handle errors?

File: vm/src/warn.rs [hunk 21]
```diff
@@ -31,19 +78,172 @@ pub fn warn(
...
+                    return Ok(false);
+                }
+            }
+        }
+    }
+
+    /* This warning wasn't found in the registry, set it. */
+    Ok(if should_set {
+        let item = vm.ctx.true_value.clone().into();
+        registry.set_item(key.as_ref(), item, vm).map(|_| true)?
```
Comment (youknowone) [inline vm/src/warn.rs:204]:
```suggestion
        let _ = registry.set_item(key.as_ref(), item, vm);  // ignore set error
        Ok(true)
```

File: vm/src/warn.rs [hunk 22]
```diff
@@ -31,19 +78,172 @@ pub fn warn(
...
+fn normalize_module(filename: PyStrRef, vm: &VirtualMachine) -> Option<PyObjectRef> {
+    let len = filename.char_len();
+
+    if len == 0 {
+        Some(vm.new_pyobj("<unknown>"))
+    } else if len >= 3 && filename.as_str().ends_with(".py") {
+        Some(vm.new_pyobj(&filename.as_str()[..len - 3]))
+    } else {
+        Some(filename.as_object().to_owned())
+    }
```
Comment (youknowone) [inline vm/src/warn.rs:219]: match works better for this kind of case

```suggestion
    let obj = match filename.char_len() {
        0 => vm.new_pyobj("<unknown>"),
        len if len >= 3 && filename.as_str().ends_with(".py") => {
            vm.new_pyobj(&filename.as_str()[..len - 3])
        }
        _ => filename.as_object().to_owned(),
    };
    Some(obj)
```

**Review by youknowone — APPROVED** [review]

**Pull request merged**